### PR TITLE
Kubernetes pod and namespace security model

### PR DESCRIPTION
**Pull request by smarterclayton**
This proposed update to docs/design/security.md includes proposals
on how to ensure containers have consistent Linux security behavior
across nodes, how containers authenticate and authorize to the master
and other components, and how secret data could be distributed to
pods to allow that authentication.

References concepts from #3910, #2030, and #2297 as well as upstream issues
around the Docker vault and Docker secrets.

Pulled together to frame the discussion from #3910 as a cross Kubernetes
concept as per @erictune

### CODE REVIEW

**Comment by smarterclayton** [comment]
Also references #4126

**Comment by erictune** [comment]
Will look at monday when I return to work.

File: docs/design/security.md [hunk 1]
```diff
@@ -1,17 +1,106 @@
...
+
+* Project Admins:
+  * allocate identity and roles within a namespace
+  * reason about organizational security within a namespace
+    * don't give a developer permissions that are not needed for role.
+    * protect files on shared storage from unnecessary cross-team access
+  * are less focused about application security
+
+* Administrators:
+  * are less focused on application security. Focused on operation system security.
```
Comment (erictune) [inline docs/design/security.md]: s/operation/operating

File: docs/design/security.md [hunk 2]
```diff
@@ -1,17 +1,106 @@
...
+
+## Proposed Design
+
+A pod runs in a *security context* under a *service account* that is defined by an administrator or project administrator, and the *secrets* a pod has access to is limited by that *service account*.
+
+
+1. The API should authenticate and authorize user actions [authn and authz](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/access.md)
+2. All infrastructure components (kubelets, kube-proxies, controllers, scheduler) should have an infrastructure user that they can authenticate with and be authorized to perform only the functions they require against the API.
+3. Most infrastructure components should use the API as a way of exchanging data and changing the system, and only the API should have access to the underlying data store (etcd)
+4. When containers run on the cluster and need to talk to other containers or the API server, they should be identified and authorized clearly as an autonomous process via a [service account](https://github.com/GoogleCloudPlatform/kubernetes/pull/2297)
```
Comment (erictune) [inline docs/design/security.md]: markdown fail

File: docs/design/security.md [hunk 3]
```diff
@@ -1,17 +1,106 @@
...
+   a.  If the user who started a long-lived process is removed from access to the cluster, the process should be able to continue without interruption
+   b.  If the user who started processes are removed from the cluster, administrators may wish to terminate their processes in bulk
+   c.  When containers run with a service account, the user that created / triggered the service account behavior must be associated to the container's action
```
Comment (erictune) [inline docs/design/security.md]: Not sure about 4b and 4c.

File: docs/design/security.md [hunk 4]
```diff
@@ -1,17 +1,106 @@
...
+   c.  Administrators should be able to limit which developers and project administrators have access to higher privilege actions
+   d.  Project administrators should be able to run pods within a namespace under different security contexts, and developers must be able to specify which of the available security contexts they may use
+   e.  Developers should be able to run their own images or images from the community and expect those images to run correctly
+   f.  Developers may need to ensure their images work within higher security requirements specified by administrators
+   g.  When available, Linux kernel user namespaces can be used to ensure b. and d. are met.
+   h.  When application developers want to share filesytem data via distributed filesystems, the Unix user ids on those filesystems must be consistent across different container processes
+6. Developers should be able to define [secrets](https://github.com/GoogleCloudPlatform/kubernetes/pull/2297) that are automatically added to the containers when pods are run
+   a.  Secrets are files injected into the container whose values should not be displayed within a pod. Examples:
```
Comment (erictune) [inline docs/design/security.md]: clarify if "should not be displayed within a pod" means "should not be displayed by GET /pods" or "should not be visible to other containers in the same pod."  I think you mean the former.

File: docs/design/security.md [hunk 5]
```diff
@@ -1,17 +1,106 @@
...
+       1. An SSH private key for git cloning remote data
+       2. A client certificate for accessing a remote system
+       3. A private key and certificate for a web server
+       4. A .kubeconfig file with embedded cert / token data for accessing the Kubernetes master
+       5. A .dockercfg file for pulling images from a protected registry
+   b.  Developers should be able to define the pod spec so that a secret lands in a specific location
+   c.  Project administrators should be able to limit developers within a namespace from viewing or modify secrets (anyone who can launch an arbitrary pod can view secrets)
+   d.  Secrets are generally not copied from one namespace to another when a developer's application definitions are copied
```
Comment (erictune) [inline docs/design/security.md]: I hope we can also support the "dev and prod instances in the same namespace using labels for separation" model.  This means "Access to secrets can be controlled via the labels on a pod".

**Comment by erictune** [comment]
minor typos, and we will need to come back to a discussion of how practical 4b and 4c are.  But LGTM.

File: docs/design/security.md (hunk 3 above)
Comment (smarterclayton) [inline docs/design/security.md]: I think it's a timeframe thing.  Ultimately someone needs to correlate ownership of resources to organizational changes, so I'd want to have it called out as a scenario somehow.

File: docs/design/security.md (hunk 5 above)
Comment (smarterclayton) [inline docs/design/security.md]: Does that mean that labels have to be access controlled - a developer who can create dev stuff can't create something that uses the label for prod stuff?  That seems like selective access to service accounts (a service account could say which users/roles can use it)?

**Comment by smarterclayton** [comment]
Comments and questions updated.

----- Original Message -----

**Comment by benmccann** [comment]
You might be interested in this recent security design for etcd. Would be interesting to get your feedback there: https://github.com/coreos/etcd/issues/2384

**Pull request merged**